addons/block_code/serialization/block_script_serialization.gd

```diff
@@ -79,24 +79,29 @@ func instantiate_block_by_name(block_name: String) -> Block:
 	return instantiate_block(block_definition)
 
 
-func get_block_definition(block_name: String) -> BlockDefinition:
+func get_block_definition(block_name: String, arguments: Dictionary) -> BlockDefinition:
 	var split := block_name.split(":", true, 1)
+	var block_definition: BlockDefinition
 
 	if len(split) > 1:
 		return _get_parameter_block_definition(split[0], split[1])
 
-	var block_definition = _get_base_block_definition(block_name)
+	block_definition = _get_base_block_definition(block_name)
+	if block_definition != null:
+		return block_definition
 
-	if block_definition == null:
-		# FIXME: This is a workaround for old-style output block references.
-		#        These were generated ahead of time using a block name that has
-		#        a "_" before the parameter name. Now, these parameter blocks
-		#        are generated on demand for any block name containing a ":".
-		#        Please remove this fallback when it is no longer necessary.
-		split = block_name.rsplit("_", true, 1)
-		return _get_parameter_block_definition(split[0], split[1])
+	var property_value = arguments.get("value", null)
+	block_definition = get_obj_property_block_definition(block_name, property_value)
+	if block_definition != null:
+		return block_definition
```

Comment:
Have you tried building this feature around `_get_parameter_block_definition`? There's a bit of overlap here. In fact they're solving the same fundamental problem: it's a type of block where we generate a block definition when we need it, based on an identifier such as a property name.

As it exists, that function only generates the one kind of value block, which is an output parameter for another block. (It needs a base block definition so it knows what output parameters are defined). So it needs extra help, but it should be feasible to extend it similarly to what you have in `get_obj_property_block_definition`, so our parameter blocks would start looking like _MyType_get:property_.

I think there's a few ways to make that work. I think ideally we'd have `_get_base_block_definition` generate a plain _MyType_get_ and _MyType_set_ block for whatever type is at hand, and we'd have something in BlockDefinition that `_get_parameter_block_definition` calls which would be like "give me a parameterized version of this block definition with foo set to bar".

I realize I'm asking for a fair bit of extra machinery here so I understand if we just want to land a bug fix, but I did have this case in mind when I was adding parameter block definitions. Sorry I didn't really document it :)

Reply:
Let me study this more.

Reply:
I try some scenario as examples: the "speed" and "floor_stop_on_slope" properties
* "Object_get:speed" will be split for `_get_parameter_block_definition()`
  - block_name: "Object_get"
  - parameter_name: "speed"
*  "Object_get:floor_stop_on_slope" will be split for `_get_parameter_block_definition()`
  - block_name: "Object_get"
  - parameter_name: "floor_stop_on_slope"

Because of the different `parameter_name`s, we have to generate different base block definitions for [`parent_out_parameters.has(parameter_name)`](https://github.com/endlessm/godot-block-coding/blob/c1193269ec0d655b97486419b45a10d78fa4c4f0/addons/block_code/serialization/block_script_serialization.gd#L117), which means the object property block's logic is different from the parameter block.

Reply:
Besides, it loses the property's variant type after close the project actually.  Because, the property's variant type is not saved in the resource of `.tscn`.  The setter block's variant type can be restored by the stored **value**.  However, getter block does not have the value information.

Reply:
Based on the limitation of the **parameter_name** and the getter's variant type, I think we cannot generate blocks via `_get_parameter_block_definition()` now.  So, I go back to the `_get_obj_property_block_definition()` in the updated commits.

Reply:
Sorry for coming in late here, but the property doesn't need to be encoded in the scene file because it's part of the object it's being applied to. It's more work, but you can call [get_property_list](https://docs.godotengine.org/en/stable/classes/class_object.html#class-object-method-get-property-list) on the target object to determine the type. I think that would be more elegant than encoding the type in the block name. It wouldn't be that hard to get the property list once at the beginning of deserializing and then find the appropriate property as the definitions are loaded.